src/Data/Conduit/Tar.hs

```diff
@@ -270,3 +245,482 @@ withEntries :: MonadThrow m
             => (Header -> ConduitM ByteString o m ())
             -> ConduitM TarChunk o m ()
 withEntries = peekForever . withEntry
+
+
+-- | Extract a tarball, similarly to `withEntries`, but instead of dealing directly with tar format,
+-- this conduit allows you to work directly on file abstractions `FileInfo`. For now support is
+-- minimal:
+--
+-- * Old v7 tar format.
+-- * ustar: POSIX 1003.1-1988 format
+-- * and only portions of GNU format:
+--   * 'L' type - long file names, but only up to 4096 chars to prevent DoS attack
+--   * other types are simply discarded
+--
+-- /Note/ - Here is a really good reference for specifics of different tar formats:
+-- <https://github.com/libarchive/libarchive/wiki/ManPageTar5>
+withFileInfo :: MonadThrow m
+             => (FileInfo -> ConduitM ByteString o m ())
+             -> ConduitM TarChunk o m ()
+withFileInfo inner = do
+    mc <- await
+    case mc of
+        Nothing -> return ()
+        Just (ChunkHeader h)
+            | headerLinkIndicator h >= 55 -> do
+                when (headerMagicVersion h == gnuTarMagicVersion) $ handleGnuTarHeader h
+                withFileInfo inner
+        Just (ChunkHeader h) -> do
+            payloadsConduit .| (inner (fileInfoFromHeader h) <* sinkNull)
+            withFileInfo inner
+        Just x@(ChunkPayload offset _bs) -> do
+            leftover x
+            throwM $ UnexpectedPayload offset
+        Just (ChunkException e) -> throwM e
+
+
+-- | Take care of custom GNU tar format.
+handleGnuTarHeader :: MonadThrow m
+                   => Header
+                   -> ConduitM TarChunk o m ()
+handleGnuTarHeader h = do
+    case headerLinkIndicator h of
+        76 -> do
+            let pSize = headerPayloadSize h
+            -- guard against names that are too long in order to prevent a DoS attack on unbounded
+            -- file names
+            unless (0 < pSize && pSize <= 4096) $
+                throwM $
+                FileTypeError (headerPayloadOffset h) 'L' $ "Filepath is too long: " ++ show pSize
+            longFileNameBuilder <- payloadsConduit .| sinkBuilder
+            let longFileName = SL.toStrict . SL.init . toLazyByteString $ longFileNameBuilder
+            mcNext <- await
+            case mcNext of
+                Just (ChunkHeader nh) -> do
+                    unless (S.isPrefixOf (fromShort (headerFileNameSuffix nh)) longFileName) $
+                        throwM $
+                        FileTypeError (headerPayloadOffset nh) 'L' $
+                        "Long filename doesn't match the original."
+                    leftover
```

Comment:
Clever, but unexpected usage of leftovers. I'd argue that it would make more sense to return the updated header from this function for extra clarity.

Reply:
Switch to `yield`, which now looks more idiomatic.

Reply:
Dammit, this idiomatic change introduced a bug, grrr. Just discovered it and now trying to understand why it is the case